common/libzkp/impl/src/proofs.rs

```diff
@@ -0,0 +1,344 @@
+    fn as_evm_proof(&self) -> &EvmProof {
+        self.proof
+            .as_evm_proof()
+            .expect("bundle proof use evm proof")
+    }
+}
+
+impl IntoEvmProof for BundleProof {
+    fn into_evm_proof(self) -> OpenVmEvmProof {
+        self.proof
+            .as_evm_proof()
+            .expect("bundle proof use evm proof")
+            .clone()
+            .into()
+    }
+}
+
+/// Trait to enable operations in metadata
+pub trait ProofMetadata: Serialize + DeserializeOwned + std::fmt::Debug {
+    type PublicInputs: MultiVersionPublicInputs;
+
+    fn pi_hash_info(&self) -> &Self::PublicInputs;
+
+    fn new_proof<P: Into<ProofEnum>>(self, proof: P, vk: Option<&[u8]>) -> WrappedProof<Self> {
+        WrappedProof {
+            metadata: self,
+            proof: proof.into(),
+            vk: vk.map(Vec::from).unwrap_or_default(),
+            git_version: short_git_version(),
+        }
+    }
+}
+
+pub trait PersistableProof: Sized {
+    /// Read and deserialize the proof.
+    fn from_json<P: AsRef<Path>>(path_proof: P) -> Result<Self>;
+    /// Serialize the proof and dumping at the given path.
+    fn dump<P: AsRef<Path>>(&self, path_proof: P) -> Result<()>;
+}
+
+/// Metadata attached to [`ChunkProof`].
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ChunkProofMetadata {
+    /// The chunk information describing the list of blocks contained within the chunk.
+    pub chunk_info: ChunkInfo,
+}
+
+impl ProofMetadata for ChunkProofMetadata {
+    type PublicInputs = ChunkInfo;
+
+    fn pi_hash_info(&self) -> &Self::PublicInputs {
+        &self.chunk_info
+    }
+}
+
+/// Metadata attached to [`BatchProof`].
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct BatchProofMetadata {
+    /// The batch information describing the list of chunks.
+    pub batch_info: BatchInfo,
+    /// The [`scroll_zkvm_types::batch::BatchHeader`]'s digest.
+    pub batch_hash: B256,
+}
+
+impl ProofMetadata for BatchProofMetadata {
+    type PublicInputs = BatchInfo;
+
+    fn pi_hash_info(&self) -> &Self::PublicInputs {
+        &self.batch_info
+    }
+}
+
+/// Metadata attached to [`BundleProof`].
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct BundleProofMetadata {
+    /// The bundle information describing the list of batches to be finalised on-chain.
+    pub bundle_info: BundleInfo,
+    /// The public-input digest for the bundle.
+    pub bundle_pi_hash: B256,
+}
+
+impl ProofMetadata for BundleProofMetadata {
+    type PublicInputs = BundleInfo;
+
+    fn pi_hash_info(&self) -> &Self::PublicInputs {
+        &self.bundle_info
+    }
+}
+
+impl<Metadata> From<&WrappedProof<Metadata>> for AggregationInput {
+    fn from(value: &WrappedProof<Metadata>) -> Self {
+        Self {
+            public_values: value.proof.public_values(),
+            commitment: ProgramCommitment::deserialize(&value.vk),
+        }
+    }
+}
+
+impl<Metadata: ProofMetadata> WrappedProof<Metadata> {
```

Comment:
in my opinion, sanity_check should not happen like this.

As long as the proof.pi == proving_task.pi (and it passes some universial check), it is a good proof.   It is proving_task building phase that it should check the consistency between proving_task.pi <-> proving_task.metadata.